subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/transaction/CompileTransaction.java

```diff
@@ -0,0 +1,302 @@
+        File headerOutput = spec.getCompileOptions().getHeaderOutputDirectory();
+        List<StashedFile> stashedFiles = new ArrayList<>();
+        for (File sourceFile : collectFilesToStash(compileOutput, annotationProcessorOutput, headerOutput)) {
+            File stashedFile = new File(stashDirectory, sourceFile.getName() + ".uniqueId" + uniqueId++);
+            moveFile(sourceFile, stashedFile);
+            stashedFiles.add(new StashedFile(sourceFile, stashedFile));
+        }
+        List<File> sourceDirectories = Stream.of(compileOutput, annotationProcessorOutput, headerOutput)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+        return new StashResult(sourceDirectories, stashedFiles);
+    }
+
+    private Set<File> collectFilesToStash(File compileOutput, @Nullable File annotationProcessorOutput, @Nullable File headerOutput) {
+        Set<File> filesToStash = new HashSet<>();
+        filesToStash.addAll(collectFilesToStash(classesToDelete, compileOutput));
+        filesToStash.addAll(collectFilesToStash(classesToDelete, annotationProcessorOutput));
+        filesToStash.addAll(collectFilesToStash(classesToDelete, headerOutput));
+        filesToStash.addAll(collectFilesToStash(resourcesToDelete.get(CLASS_OUTPUT), compileOutput));
+        // If the client has not set a location for SOURCE_OUTPUT, javac outputs those files to the CLASS_OUTPUT directory, so delete that instead.
+        filesToStash.addAll(collectFilesToStash(resourcesToDelete.get(SOURCE_OUTPUT), MoreObjects.firstNonNull(annotationProcessorOutput, compileOutput)));
+        // In the same situation with NATIVE_HEADER_OUTPUT, javac just NPEs.  Don't bother.
+        filesToStash.addAll(collectFilesToStash(resourcesToDelete.get(NATIVE_HEADER_OUTPUT), headerOutput));
+        return filesToStash;
+    }
+
+    private Set<File> collectFilesToStash(PatternSet patternSet, File sourceDirectory) {
+        if (patternSet != null && !patternSet.isEmpty() && sourceDirectory != null && sourceDirectory.exists()) {
+            return fileOperations.fileTree(sourceDirectory).matching(patternSet).getFiles();
+        }
+        return Collections.emptySet();
+    }
+
+    private void deletePotentiallyEmptyDirectories(StashResult stashResult) {
+        Set<File> potentiallyEmptyFolders = stashResult.stashedFiles.stream()
+            .map(file -> file.sourceFile.getParentFile())
+            .collect(Collectors.toSet());
+        StaleOutputCleaner.cleanEmptyOutputDirectories(deleter, potentiallyEmptyFolders, stashResult.sourceDirectories);
+    }
+
+    private void moveCompileOutputToOriginalFolders(List<StagedOutput> stagedOutputs) {
+        stagedOutputs.forEach(output -> moveAllFilesFromDirectoryTo(output.stagingDirectory, output.sourceDirectory));
+    }
+
+    private void moveAllFilesFromDirectoryTo(File sourceDirectory, File destinationDirectory) {
```

Comment:
Just because I am a fan of "data classes". So that data is just a simple POJO, and some function/methods is applied to them, but they don't have any special logic.

Reply:
I really don't like data classes since they work against a few of the OOP principles.